Userland/Libraries/LibGL/SoftwareGLContext.cpp

```diff
@@ -74,6 +75,43 @@ SoftwareGLContext::SoftwareGLContext(Gfx::Bitmap& frontbuffer)
     light0.diffuse_intensity = { 1.0f, 1.0f, 1.0f, 1.0f };
     light0.specular_intensity = { 1.0f, 1.0f, 1.0f, 1.0f };
     m_light_state_is_dirty = true;
+
+    // Build gl extension string
+    Vector<StringView> extensions;
+
+    // FIXME: npot texture support became a required core feature starting with OpenGL 2.0 (https://www.khronos.org/opengl/wiki/NPOT_Texture)
+    // Ideally we would verify if the selected device adheres to the requested OpenGL context version before context creation
+    // and refuse to create a context if it doesn't.
+    if (m_device_info.supports_npot_textures)
+        extensions.append("GL_ARB_texture_non_power_of_two");
+
+    if (m_device_info.num_texture_units > 1)
+        extensions.append("GL_ARB_multitexture");
+
+    m_client_side_texture_coord_array_enabled.resize(m_device_info.num_texture_units);
+    m_client_tex_coord_pointer.resize(m_device_info.num_texture_units);
+    m_current_vertex_tex_coord.resize(m_device_info.num_texture_units);
+    for (auto& tex_coord : m_current_vertex_tex_coord)
+        tex_coord = { 0.0f, 0.0f, 0.0f, 1.0f };
+
+    m_texture_coordinate_generation.resize(m_device_info.num_texture_units);
+    for (auto& texture_coordinate_generation : m_texture_coordinate_generation) {
+        texture_coordinate_generation[0].object_plane_coefficients = { 1.0f, 0.0f, 0.0f, 0.0f };
+        texture_coordinate_generation[0].eye_plane_coefficients = { 1.0f, 0.0f, 0.0f, 0.0f };
+        texture_coordinate_generation[1].object_plane_coefficients = { 0.0f, 1.0f, 0.0f, 0.0f };
+        texture_coordinate_generation[1].eye_plane_coefficients = { 0.0f, 1.0f, 0.0f, 0.0f };
+    }
+
+    StringBuilder builder;
+    bool is_first = true;
+    for (auto const& extension : extensions) {
+        if (!is_first)
+            builder.append(" ");
+        is_first = false;
+
+        builder.append(extension);
+    }
+    m_extensions = builder.to_string();
```

Comment:
Usually, extensions are only requested once. Maybe this is better suited in a separate function instead of in the constructor, and build the string on-demand?

Reply:
Sure, putting that in a separate function makes a lot of sense.
I think building the string in the constructor is a good idea though since we are doing a lot of setup and verification work there anyway.

Reply:
Features/verification in the constructor is fine, but I don't see the practical use of keeping the extension string in memory for the entire duration of the context?

Reply:
We only return a pointer to it, so it needs to stay alive after returning from the `glGetString()` call.

Reply:
Ah I missed that. Makes total sense! Resolving :-)